adam/curriculum/preposition_curriculum.py

```diff
@@ -0,0 +1,726 @@
+    object_variable,
+)
+
+_CHOOSER = RandomChooser.for_seed(0)
+
+
+def _make_background(
+    salient: Iterable[TemplateObjectVariable],
+    all_objects: Iterable[TemplateObjectVariable],
+) -> Iterable[TemplateObjectVariable]:
+    return immutableset(object_ for object_ in all_objects if object_ not in salient)
+
+
+def _on_templates(
+    figure: TemplateObjectVariable,
+    ground: TemplateObjectVariable,
+    background: Iterable[TemplateObjectVariable],
+) -> Phase1SituationTemplate:
+    return Phase1SituationTemplate(
+        f"preposition-training-{figure.handle}-on-{ground.handle}",
+        salient_object_variables=[figure, ground],
+        background_object_variables=background,
+        asserted_always_relations=[on(figure, ground)],
+    )
+
+
+def _beside_template(
+    figure: TemplateObjectVariable,
+    ground: TemplateObjectVariable,
+    background: Iterable[TemplateObjectVariable],
+    *,
+    is_right: bool,
+) -> Phase1SituationTemplate:
+    direction_str = "right" if is_right else "left"
+    return Phase1SituationTemplate(
+        f"preposition-training-{figure.handle}-beside-{ground.handle}-{direction_str}",
+        salient_object_variables=[figure, ground],
+        background_object_variables=background,
+        asserted_always_relations=[
+            Relation(
+                IN_REGION,
+                figure,
+                Region(
+                    ground,
+                    distance=PROXIMAL,
+                    direction=Direction(
+                        positive=is_right,
+                        relative_to_axis=HorizontalAxisOfObject(ground, index=0),
+                    ),
+                ),
+            )
+        ],
+    )
+
+
+def _over_template(
+    figure: TemplateObjectVariable,
+    ground: TemplateObjectVariable,
+    background: Iterable[TemplateObjectVariable],
+) -> Phase1SituationTemplate:
+    return Phase1SituationTemplate(
+        f"preposition-training-{figure.handle}-over-{ground.handle}",
+        salient_object_variables=[figure],
+        background_object_variables=background,
+        asserted_always_relations=[strictly_above(ground, figure)],
+        constraining_relations=[bigger_than(ground, figure)],
+    )
+
+
+def _under_template(
+    figure: TemplateObjectVariable,
+    ground: TemplateObjectVariable,
+    background: Iterable[TemplateObjectVariable],
+) -> Phase1SituationTemplate:
+    return Phase1SituationTemplate(
+        f"preposition-training-{figure.handle}-under-{ground.handle}",
+        salient_object_variables=[figure],
+        background_object_variables=background,
+        asserted_always_relations=[strictly_above(figure, ground)],
+    )
+
+
+def _in_template(
+    figure: TemplateObjectVariable,
+    ground: TemplateObjectVariable,
+    background: Iterable[TemplateObjectVariable],
+) -> Phase1SituationTemplate:
+    return Phase1SituationTemplate(
+        f"preposition-training-{figure.handle}-in-{ground.handle}",
```

Comment:
Do we need to assert the ground is `HOLLOW`?

Reply:
That's done inside the curriculum constructors as a `required_property` of ground.